src/libsync/discovery.cpp

```diff
@@ -1566,6 +1570,34 @@ bool ProcessDirectoryJob::checkPermissions(const OCC::SyncFileItemPtr &item)
     return true;
 }
 
+bool ProcessDirectoryJob::isAnyParentBeingRestored(const QString &file) const
+{
+    for (const auto &directoryNameToRestore : _discoveryData->_directoryNamesToRestoreOnPropagation) {
+        if (file.startsWith(QString(directoryNameToRestore + QLatin1Char('/')))) {
+            qCWarning(lcDisco) << "File" << file << " is within the tree that's being restored" << directoryNameToRestore;
+            return true;
+        }
+    }
+    return false;
+}
+
+bool ProcessDirectoryJob::isRenameAndIsAllowed(const QString &originalPath, const QString &targetPath) const
+{
+    if (!(originalPath.startsWith(_currentFolder._original)
+        && originalPath.lastIndexOf('/') == _currentFolder._original.size())) {
+        // early-out if it is a move, not a rename
```

Comment:
I do not understand the distinction you do between a move and a rename
is a move moving the folder to another place without changing its name ?
what about moving the folder to another parent folder and changing the name at the same time ?

Reply:
@mgallien Thanks for this comment. I have double-checked.

Move - is the change of parent. E.G. - "**A/b/c/**file.txt" -> "**A/a1/**file.txt" - the location changes in a structure.
Rename - is the change of name. E.G. - "A/b/c/**file**.txt" -> "A/b/c/**file-enamed**.txt" - the location changes in a structure.

We have 2 main points:
- A rename can be allowed when the move is not, so, if it is not a move - we must check for rename permission
- If it is a move, we do not check for nested renames, as, in this case, we must reupload all files anyways

Your point about moving the folder to another parent and changing the name at the same time - this is possible, though is an edge case which is unlikely. Since it is an artificial scenario, there is no reliable way to detect both moves and renames such that we can decide which of the two we allow - the best way is to prevent the rename of an item if it has been moved while the move was forbidden. I hope that makes sense?

These condition does not prevent normal rename (without also moving at the same time) if rename permissions are present.